proposals/0000-instance-foralls.rst

```diff
@@ -0,0 +1,127 @@
+
+On type signatures, users can write ``forall ...`` to explicitly bind type/kind variables.
+This allows users to give the type variables explicit kinds. However, not all constructs that
+bind type variables allow the user to write them explicitly. This proposal seeks to remedy this
+by enhancing GHC's concrete syntax to permit ``forall ...`` expressions wherever type variables
+may be bound.
+
+Specifically, this proposal adds the ability to write a ``forall ...`` clause on the following
+constructs:
+
+* data/newtype instances (including associated types)
+* type instances (including associated types)
+* closed type family equations
+* ``RULES``
+
+In contrast, the following constructs already allow user-written kinds for type variables:
+
+* type/data families (including associated ones): The type variables listed are type variable
+  binders, not occurrences.
+* type synonyms: The type variables listed are binders.
+* data/newtypes: The type variables listed are binders, in both Haskell98 and GADT notation.
+* classes: The type variables listed are binders.
+* class instances: The user can write an optional ``forall ...`` after the word ``instance``.
+* standalone-deriving instances: Ditto.
+* values: The user can bring scoped type variables into scope via a type signature.
+* type signatures: The user can write an optional ``forall ...``.
+* pattern signatures: The user can write separate optional ``forall ...``\s for both universal
+  and existential type variables.
+* ``SPECIALIZE`` pragmas: The user can write an optional ``forall ...`` in the type.
+* ``SPECIALIZE instance`` pragmas: The user can write an optional ``forall ...`` after the word ``instance``.
+
+I believe the list above is a full accounting of places in the GHC concrete syntax where
+type variables may appear. Do comment if I've missed anything.
+
+Motivation
+------------
+Given the evolution in GHC's kind system (``TypeInType`` in particular),
+it is no longer possible always to *infer* the kind
+of a type variable. For example, a type variable ``a`` might be used only in a place where it is expected
+to have kind ``F b`` for some non-injective type family ``F``. We cannot confidently choose ``a`` to
+have kind ``F b`` in this circumstance. Similarly, GHC now supports *higher-rank kinds*, where a type
+variable's kind might be a polytype. These, too, cannot be inferred.
+
+Note that a use of a type variable in a kind signature ``(a :: F b)`` does not necessarily give
+the declared kind for ``a``.
+
+Due to the inability to always infer kinds, the user may need to specify a kind manually. This proposal
+allows the user to do this in all syntactic scenarios.
+
+Proposed Change Specification
+-----------------------------
+
+1. Permit ``forall ...`` to bind type variables after the word ``instance`` in a ``data``,
+   ``newtype``, or ``type`` instance. For an associated type, the ``forall`` would appear after the word
+   ``data``, ``newtype``, or ``type``.
```

Comment:
@Ericson2314 I'm not sure what that is meant to mean. Do you mean `Type` instead of `type`? I assume so. Then it looks like you're giving a top-level kind signature on an associated type *instance*, which is not part of #54. If we drop the first line within the instance, then it all looks good to me. But then I don't see how it interacts with this current PR. Sorry if I'm being dense!